### ISSUE TITLE
GetVideoFromGallery crashes iOS app if looping live picture is selected

### ISSUE THREAD

**Issue by jmcgee412** [yasirkula/UnityNativeGallery#159]
On iOS, you have the option of creating a Live Photo...and after you've created the Live Photo, there are a couple of options you can do with it. One of the options is to make it a "Looping" Photo

On your iPhone, when looking at your Albums, you have the option of viewing them by Media Types (videos, live photos, screenshots, etc.)...If you select Videos, then your Live Photos are not shown...However, when using the plugin in my app, using the GetVideoFromGallery method, Live Photos that have been made to Loop ARE shown, along with the other videos on the phone. And if the user selects one of these "Looping" Live Photos, then it crashes the app

This is the error I get on Xcode:

2020-12-30 00:39:48.303965-0800 VuforiaVideoPlayerTest[1264:887710] Picked an image
2020-12-30 00:39:48.538577-0800 VuforiaVideoPlayerTest[1264:887748] [default] [ERROR] Could not create a bookmark: NSError: Cocoa 257 "The file couldn’t be opened because you don’t have permission to view it." }
2020-12-30 00:39:48.583162-0800 VuforiaVideoPlayerTest[1264:887747] [Technique] ARWorldTrackingTechnique <0x138956e10>: World tracking performance is being affected by resource constraints [1]
2020-12-30 00:39:49.960669-0800 VuforiaVideoPlayerTest[1264:887750] Error copying file type public.image. Error: Error Domain=NSItemProviderErrorDomain Code=-1000 "Cannot load representation of type public.image" UserInfo={NSLocalizedDescription=Cannot load representation of type public.image, NSUnderlyingError=0x281ed4330 {Error Domain=NSCocoaErrorDomain Code=4101 "Couldn’t communicate with a helper application." UserInfo={NSUnderlyingError=0x281db9ef0 {Error Domain=PFPAssetRequestErrorDomain Code=0 "(null)"}}}}
2020-12-30 00:39:49.960757-0800 VuforiaVideoPlayerTest[1264:887750] Error getting the picked image's path: Error Domain=NSItemProviderErrorDomain Code=-1000 "Cannot load representation of type public.image" UserInfo={NSLocalizedDescription=Cannot load representation of type public.image, NSUnderlyingError=0x281ed4330 {Error Domain=NSCocoaErrorDomain Code=4101 "Couldn’t communicate with a helper application." UserInfo={NSUnderlyingError=0x281db9ef0 {Error Domain=PFPAssetRequestErrorDomain Code=0 "(null)"}}}}
2020-12-30 00:39:51.323294-0800 VuforiaVideoPlayerTest[1264:887748] [claims] Upload preparation for claim C1DD9996-FD76-4530-ADA9-478F5DBBA70E completed with error: Error Domain=NSCocoaErrorDomain Code=260 "The file “version=1&uuid=9937F2AB-A026-4D90-BD19-ACC4521CC949&mode=current.gif” couldn’t be opened because there is no such file." UserInfo={NSURL=file:///private/var/mobile/Containers/Shared/AppGroup/72519A3E-6DAC-4B19-846D-7BD3B3E90053/File%20Provider%20Storage/photospicker/version=1&uuid=9937F2AB-A026-4D90-BD19-ACC4521CC949&mode=current.gif, NSFilePath=/private/var/mobile/Containers/Shared/AppGroup/72519A3E-6DAC-4B19-846D-7BD3B3E90053/File Provider Storage/photospicker/version=1&uuid=9937F2AB-A026-4D90-BD19-ACC4521CC949&mode=current.gif, NSUnderlyingError=0x281dbac70 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}
2020-12-30 00:39:51.323693-0800 VuforiaVideoPlayerTest[1264:887748] Error generating UIImage from picked image: Error Domain=NSItemProviderErrorDomain Code=-1000 "Cannot load representation of type com.compuserve.gif" UserInfo={NSLocalizedDescription=Cannot load representation of type com.compuserve.gif, NSUnderlyingError=0x281dba790 {Error Domain=NSCocoaErrorDomain Code=260 "The file “version=1&uuid=9937F2AB-A026-4D90-BD19-ACC4521CC949&mode=current.gif” couldn’t be opened because there is no such file." UserInfo={NSURL=file:///private/var/mobile/Containers/Shared/AppGroup/72519A3E-6DAC-4B19-846D-7BD3B3E90053/File%20Provider%20Storage/photospicker/version=1&uuid=9937F2AB-A026-4D90-BD19-ACC4521CC949&mode=current.gif, NSFilePath=/private/var/mobile/Containers/Shared/AppGroup/72519A3E-6DAC-4B19-846D-7BD3B3E90053/File Provider Storage/photospicker/version=1&uuid=9937F2AB-A026-4D90-BD19-ACC4521CC949&mode=current.gif, NSUnderlyingError=0x281dbac70 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}}}
2020-12-30 00:39:51.347133-0800 VuforiaVideoPlayerTest[1264:887748] Uncaught exception: NSInvalidArgumentException: *** -[__NSArrayM insertObject:atIndex:]: object cannot be nil
(
	0   CoreFoundation                      0x000000019f378880 935533F2-35EE-314E-A760-E74521F68435 + 1202304
	1   libobjc.A.dylib                     0x00000001b42e7c50 objc_exception_throw + 60
	2   CoreFoundation                      0x000000019f3e8e1c 935533F2-35EE-314E-A760-E74521F68435 + 1662492
	3   CoreFoundation                      0x000000019f3e6300 935533F2-35EE-314E-A760-E74521F68435 + 1651456
	4   CoreFoundation                      0x000000019f255b90 935533F2-35EE-314E-A760-E74521F68435 + 11152
	5   UnityFramework                      0x000000010534aeb4 __42+[UNativeGallery picker:didFinishPicking:]_block_invoke_2 + 348
	6   Foundation                          0x00000001a06ea37c 50B3154E-8C73-3E17-A162-C30DA82424CD + 1401724
	7   libdispatch.dylib                   0x000000019ef6b24c 0F3905E0-6720-3D08-9AE0-A43A8DA244C0 + 8780
	8   libdispatch.dylib                   0x000000019ef6cdb0 0F3905E0-6720-3D08-9AE0-A43A8DA244C0 + 15792
	9   libdispatch.dylib                   0x000000019ef7410c 0F3905E0-6720-3D08-9AE0-A43A8DA244C0 + 45324
	10  libdispatch.dylib                   0x000000019ef74c5c 0F3905E0-6720-3D08-9AE0-A43A8DA244C0 + 48220
	11  libdispatch.dylib                   0x000000019ef7ed78 0F3905E0-6720-3D08-9AE0-A43A8DA244C0 + 89464
	12  libsystem_pthread.dylib             0x00000001eabc3804 _pthread_wqthread + 276
	13  libsystem_pthread.dylib             0x00000001eabca75c start_wqthread + 8
)
2020-12-30 00:39:51.348091-0800 VuforiaVideoPlayerTest[1264:887748] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSArrayM insertObject:atIndex:]: object cannot be nil'
*** First throw call stack:
(0x19f37886c 0x1b42e7c50 0x19f3e8e1c 0x19f3e6300 0x19f255b90 0x10534aeb4 0x1a06ea37c 0x19ef6b24c 0x19ef6cdb0 0x19ef7410c 0x19ef74c5c 0x19ef7ed78 0x1eabc3804 0x1eabca75c)
libc++abi.dylib: terminating with uncaught exception of type NSException
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSArrayM insertObject:atIndex:]: object cannot be nil'
terminating with uncaught exception of type NSException
(lldb) 

I'm not 100% sure...but I think the Looping Video is stored with a GIF extension, and the storage or processing of the GIF in the temporary cached path is causing the conflict

I'm wondering if there is a way to have these files excluded from Gallery selection so the user does not have the option to choose them and crash the app. It's interesting how they do not show up when looking at the Gallery on your phone and categorizing by Video Media Types, but they do show up in the Gallery selection when using the plugin and using the video picker

To replicate this, you'll want to take a Live Photo on an iPhone. After you've taken the photo, view it. You should see a label in the upper left-hand corner of the picture that says "Live"...click that, and then slide up. You should see options for "Effects". Now choose "Loop"...Now you can test it. You should see that if you go to the Photos app on your iPhone, scroll to Media Types, and sort by Videos...it will not be there. But then if you create an app with the plugin and use the GetVideoFromGallery method, it will be displayed as one of the options. If you select it, then the app will crash

Using:
Native Gallery for Android & iOS version 1.6.1
Unity 2020.1.2f1
iPhone SE 2020 Edition using iOS 14.3
Xcode 12.2
MacOS version 10.15.7

**Comment by yasirkula** [comment]
Thank you for the detailed Issue. This is disturbing news because there are 2 rules that Apple's PHPickerViewController breaks here:

1. I only ask for videos, so it shouldn't display live photos:

https://github.com/yasirkula/UnityNativeGallery/blob/0bf62249d8c70df3442401adfcc48842098254af/Plugins/NativeGallery/iOS/NativeGallery.mm#L554-L563

2. To get the "_Error generating UIImage from picked image_" error, the following must be true:

https://github.com/yasirkula/UnityNativeGallery/blob/0bf62249d8c70df3442401adfcc48842098254af/Plugins/NativeGallery/iOS/NativeGallery.mm#L819

And the following:

https://github.com/yasirkula/UnityNativeGallery/blob/0bf62249d8c70df3442401adfcc48842098254af/Plugins/NativeGallery/iOS/NativeGallery.mm#L874

Even though both of these conditions somehow return true, the following function fails:

https://github.com/yasirkula/UnityNativeGallery/blob/0bf62249d8c70df3442401adfcc48842098254af/Plugins/NativeGallery/iOS/NativeGallery.mm#L878

You can see that loadObjectOfClass fails despite canLoadObjectOfClass returning true. **Considering all of these facts, I'd say that this looping live photo Issue is an iOS bug.**

To fix the crash, at least, please try changing the following line to `[resultPaths addObject:( resultPath != nil ? resultPath : @"" )];` and let me know if it works:

https://github.com/yasirkula/UnityNativeGallery/blob/0bf62249d8c70df3442401adfcc48842098254af/Plugins/NativeGallery/iOS/NativeGallery.mm#L893

If I had an iOS 14 device to reproduce the issue and verify that it doesn't occur on a single device, I'd report this issue on [official forums](https://developer.apple.com/forums/tags/photokit) but I don't have one so I can't.

**Comment by yasirkula** [comment]
With the latest update, at least the crash should be fixed. Can you give it a try when you get the chance @jmcgee412 ?

**Comment by jmcgee412** [comment]
First, I'd like to say to say thanks for creating such awesome, well-made plugins for us. And thanks for also being so responsive. You're a saint 

I gave the update a test run, and it no longer crashes...Thank you for making that update! I also think you're right about this being a bug on Apple's end, especially because they seem to have the Media Types filter correctly on their own Photos App, but it doesn't seem to filter correctly with their PHPickerfilter method options

I submitted the bug with Apple, we'll see if they get to it

Thanks again!
